<a href="https://colab.research.google.com/github/AnnaBo1/python_for_DA/blob/main/hw_1_3_3_Comprehensions.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Домашнє завдання до теми "Comprehensions в Python"

Для наступних завдань у нас є наступний список:

In [1]:
numbers = [42, 51, 29, 2, 13, 44, 80, 66, 16]

1. Використовуючи цикл `for` cтворіть новий список `numbers_squared`, де кожен елемент списку `numbers` є домноженим на 2.

Нагадаю, що фактично коли ми з допомогою циклу отримуємо новий список, то ми спочатку створюємо пустий список, а потім додаємо в нього новий елемент на кожній ітерації.

Новий елемент буде в даному випадку результатом домножання чергового елемента вихідного списку `numbers` на 2.

In [2]:
numbers_squared = []

In [4]:
for number in numbers :
    numbers_squared.append(number * 2) #за допомогою методу append() додає результат до списку numbers_squared

In [5]:
numbers_squared

[84, 102, 58, 4, 26, 88, 160, 132, 32, 84, 102, 58, 4, 26, 88, 160, 132, 32]

Очікуваний результат: `[84, 102, 58, 4, 26, 88, 160, 132, 32]`

2. Використовуючи `list comprehension` cтворіть новий список `numbers_squared_lh`, де кожен елемент списку `numbers` є домноженим 2.
Так, операція, яку ми викнуємо над кожним елементом вихідного списку, та сама, що і в попередньому завданні, але інший спосіб отримання результату.

In [7]:
numbers_squared_lh = [number * 2 for number in numbers]

In [8]:
numbers_squared_lh

[84, 102, 58, 4, 26, 88, 160, 132, 32]

3. У нас вже є знайдене середнє значення за списком - воно визначене в змінній `numbers_mean`.

  Використовуючи цикл, створіть новий список `numbers_gt_mean`, який порівнює кожен елемент з цим середнім значенням за списком за наступним правилом: "якщо цей елемент спиcку `numbers` є більшим за `numbers_mean`, то запишемо 1, інакше 0". Виведіть новий список на екран.

In [13]:
numbers = [42, 51, 29, 2, 13, 44, 80, 66, 16]
numbers_mean = 38.11

In [14]:
numbers_gt_mean = []

In [15]:
for number in numbers:
    if number > numbers_mean:
        numbers_gt_mean.append(1) #append додає
    else:
        numbers_gt_mean.append(0)

In [16]:
numbers_gt_mean

[1, 1, 0, 0, 0, 1, 1, 1, 0]

Очікуваний результат: `[1, 1, 0, 0, 0, 1, 1, 1, 0]`

4. Отримайте список за тим самим правилом, що в завданні 3, тільки тепер з list comprehension. Результат запишіть у змінну `numbers_gt_mean_lc`

In [17]:
numbers = [42, 51, 29, 2, 13, 44, 80, 66, 16]
numbers_mean = 38.11

In [18]:
numbers_gt_mean_lc = [1 if number > numbers_mean else 0 for number in numbers ] # for in не писати на самому початку інакше видасть помилку

In [20]:
numbers_gt_mean_lc

[1, 1, 0, 0, 0, 1, 1, 1, 0]

5. Заданий список `long`. З допомогою циклу створіть словник `n_repeats`, який містить інформацію "скільки разів повторюється кожен унікальний елемент списку `long`".
Наприклад:
```
long = [1, 1, 2, 2, 3]
# тоді `n_repeats` містить {1: 2, 2: 2, 3: 1}
```

Аби для кожного унікального елементу знайти, скільки разів він повторюється, виористайте метод списку `list.count(<елемент>)`.
Виведіть `n_repeats` на екран.

**Підказки**:
1. Для початку може бути зручно отримати набір унікальних елементів списку.
2. Порядок ключів в словнику не має значення! Тобто ключі в словнику не обовʼязково мають бути в тому ж порядку, як наведено в прикладі результату. Але кількості зустрічань для кожного елемента мають співпадати з наведеними.

In [44]:
long = [75, 39, 95, 1, 91, 43, 97, 24, 86, 75, 92, 91, 15, 15, 75, 95, 54, 29, 55, 98]

In [45]:
n_repeats = {}

In [46]:
unique = set(long) #набір унікальних елементів зі списку лонг

In [47]:
unique

{1, 15, 24, 29, 39, 43, 54, 55, 75, 86, 91, 92, 95, 97, 98}

In [48]:
for element in unique: #рахує кількість повторів за допомогою for
    n_repeats[element] = long.count(element)

In [49]:
n_repeats

{1: 1,
 97: 1,
 98: 1,
 39: 1,
 75: 3,
 43: 1,
 15: 2,
 55: 1,
 86: 1,
 54: 1,
 24: 1,
 91: 2,
 92: 1,
 29: 1,
 95: 2}

Очікуваний результат:
```
{1: 1, 97: 1, 98: 1, 39: 1, 75: 3, 43: 1, 15: 2, 55: 1, 86: 1, 54: 1, 24: 1, 91: 2, 92: 1, 29: 1, 95: 2}

```

6. Виконайте ту саму операцію, що і в попередньому заваднні, але використовуючи `dict comprehension`. Результат запишіть у змінну `n_repeats_dh` та результат виведіть на екран.

In [53]:
long = [75, 39, 95, 1, 91, 43, 97, 24, 86, 75, 92, 91, 15, 15, 75, 95, 54, 29, 55, 98]

In [55]:
n_repeats_dh = {element: long.count(element) for element in set(long)}

In [56]:
n_repeats_dh

{1: 1,
 97: 1,
 98: 1,
 39: 1,
 75: 3,
 43: 1,
 15: 2,
 55: 1,
 86: 1,
 54: 1,
 24: 1,
 91: 2,
 92: 1,
 29: 1,
 95: 2}

7. Знайдіть три найменші **унікальні** елементи в списку `long` та виведіть їх на екран.

In [116]:
long = [75, 39, 95, 1, 91, 43, 97, 24, 86, 75, 92, 91, 15, 15, 75, 95, 54, 29, 55, 98]

In [117]:
unique = set(long)

In [118]:
unique

{1, 15, 24, 29, 39, 43, 54, 55, 75, 86, 91, 92, 95, 97, 98}

In [119]:
p = list(unique)

In [120]:
p.sort()

In [121]:
p

[1, 15, 24, 29, 39, 43, 54, 55, 75, 86, 91, 92, 95, 97, 98]

In [122]:
p[:3]


[1, 15, 24]

Очікувана відповідь: [1, 15, 24]